hips/hip-00xx-sequencing.md

```diff
@@ -0,0 +1,192 @@
+layer1:    [foo]
+             ||
+             ||   [someChartDependency]
+             ||     ||
+             \/     ||
+layer2:    [bar]    ||
+              \\   //
+               \/ \/
+layer3:       [fizz]
+```
+
+This approach is prone to circular dependencies. During the templating phase, Helm will have logic to detect, and report any circular dependencies found in the chart templates.
+
+### Proposal 2: Weighted Resouces (without hooks)
+
+Expand on the the weight capabilities of hooks, but allow them to be defined outside of the context of hooks for the chart. Add a new annotation `helm.sh/weight` that helm then uses to order when resources are deployed based on the same weighting rules that hooks use today.
+
+#### Example:
+```yaml
+# resource 1
+metadata:
+  name: foo
+  annotations:
+    helm.sh/weight: "-2"
+---
+# resource 2
+metadata:
+  name: bar
+  annotations:
+    helm.sh/weight: "-1"
+---
+# resource 3
+metadata:
+  name: fizz
+  annotations:
+    helm.sh/weight: 0
+```
+In this example, Helm would be responsible for resolving the annotations on these three resources and deploy all resources in the following order:
+
+```
+-2:    [foo]
+        ||
+        \/
+-1:    [bar]
+        ||
+        \/
+0:     [fizz]
+```
+
+### Readiness
+
+In order to enforce sequencing, new `helm.sh/resource-ready` annotation would be used to determine when a resource is ready, allowing helm to proceed deploying the next group of resources, or failing a deployment. Helm would query, using jsonpath syntax, status fields of the annotated resource. Some native kubernetes resources that have stable APIs e.g `v1` resources such as `Pod` and `Deployment`, would have default queries which can be overriden. In Helm cannot determine how to check a resource's readiness when it should, it will do nothing and log this.
+
+#### Example
+```yaml
+kind: Job
+metadata:
+  name: barz
+  annotations:
+    helm.sh/resource-ready: "status.succeeded==1"
+status:
+  succeeded: 1
+```
+
+#### Sequencing order
+
+Resources with sequencing annotations would be deployed first followed by resources without. If a resource has a `helm.sh/depends-on/chart` annotation, all resources of the referred subchart would be deployed and checked for readiness before deploying the dependant resource. Only resources that Helm can determine their readiness for will be checked. Chart authors would need to annotate their chart resources accordingly. Helm will use default readiness checks for subcharts that do have their templates annotated.
+
+Helm would scope each subchart layer annotation names using a delimiter e.g `someChartDependency#mylayer` to avoid any name collisions. This is an internal implementation detail rather than feature chart authors or operators would need to know.
+
+`helm template` would print all resources in the order they would be deployed. Groups of resources in a layer would be delimited using a `## Layer: <name>` comment indicate the beginning of each layer
```

Comment:
The `#` at the front denotes a comment. I'm not sure the k8s yaml parser, that we currently use and is supported by the k8s project, provides a means to parse and use data in comments. This needs to be investigated before this possibility could be considered. See https://github.com/kubernetes-sigs/yaml

Reply:
For reference, the `## Layer: <name>` would be above/below `# Source:`. Maybe there is no need for 2 `#`. We can have it be `# Layer: <name>`.

```yaml
---
# Source: foo/templates/tests/test-connection.yaml
apiVersion: v1
kind: Pod
metadata:
```